### PR TITLE
Fold arithmetic identities in GVN

### PR DESCRIPTION
**Pull request by cjgillot**
Extracted from https://github.com/rust-lang/rust/pull/111344

This PR implements a few arithmetic folds for unary and binary operations.
This should take care of the missed optimizations introduced by https://github.com/rust-lang/rust/pull/116012.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 1]
```diff
@@ -775,14 +784,26 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
                 Value::Cast { kind, value, from, to }
             }
             Rvalue::BinaryOp(op, box (ref mut lhs, ref mut rhs)) => {
+                let ty = lhs.ty(self.local_decls, self.tcx);
                 let lhs = self.simplify_operand(lhs, location);
                 let rhs = self.simplify_operand(rhs, location);
-                Value::BinaryOp(op, lhs?, rhs?)
+                let lhs = lhs?;
+                let rhs = rhs?;
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/gvn.rs:788]: you could do the `?` op directly on the method calls above

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 2]
```diff
@@ -775,14 +784,26 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
...
+                }
+                Value::BinaryOp(op, lhs, rhs)
             }
             Rvalue::CheckedBinaryOp(op, box (ref mut lhs, ref mut rhs)) => {
+                let ty = lhs.ty(self.local_decls, self.tcx);
                 let lhs = self.simplify_operand(lhs, location);
                 let rhs = self.simplify_operand(rhs, location);
-                Value::CheckedBinaryOp(op, lhs?, rhs?)
+                let lhs = lhs?;
+                let rhs = rhs?;
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/gvn.rs:799]: same

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 3]
```diff
@@ -884,6 +905,92 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
...
+
+        let as_bits = |value| {
+            let constant = self.evaluated[value].as_ref()?;
+            let scalar = self.ecx.read_scalar(constant).ok()?;
+            scalar.to_bits(constant.layout.size).ok()
+        };
+
+        // Represent the values as `Ok(bits)` or `Err(VnIndex)`.
+        let a = as_bits(lhs).ok_or(lhs);
+        let b = as_bits(rhs).ok_or(rhs);
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/gvn.rs:957]: Use `Either` instead of `Result`?

File: compiler/rustc_mir_transform/src/gvn.rs [hunk 4]
```diff
@@ -884,6 +905,92 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
...
+            // Sub/Xor with itself.
+            (BinOp::Sub | BinOp::BitXor, a, b) if a == b => {
+                self.insert_scalar(Scalar::from_uint(0u128, layout.size), lhs_ty)
+            }
+            // Comparison:
+            // - if both operands can be computed as bits, just compare the bits;
+            // - if we proved that both operands have the same value, we can insert true/false;
+            // - otherwise, do nothing, as we do not try to prove inequality.
+            (BinOp::Eq, a, b) if (a.is_ok() && b.is_ok()) || a == b => self.insert_bool(a == b),
+            (BinOp::Ne, a, b) if (a.is_ok() && b.is_ok()) || a == b => self.insert_bool(a != b),
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/gvn.rs:1005]: why does it need to be `ok` for equality? wouldn't it also hold if they are the same `VnIndex`?

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_mir_transform/src/gvn.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/gvn.rs:800]: Added a comment explaining why it is done this way.

File: compiler/rustc_mir_transform/src/gvn.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/gvn.rs:813]: same

**Comment by cjgillot** [comment]
Blessed tests after rebasing.
@bors r=oli-obk

**Comment by cjgillot** [comment]
Perf effect is likely.
@bors rollup=never

**Comment by ChrisDenton** [comment]
Though this is a perf only fix, would T-compiler consider backporting?

cc @mati865

**Comment by apiraino** [comment]
Beta backport declined as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-02-01/near/419267038), also [comment](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-02-01/near/419323057) from PR author.

backporting this it's a bit too risky

@rustbot label -beta-nominated

**Pull request merged**